edge-view for EVE

naiming-zededa opened the pull request:
Signed-off-by: Naiming Shen <naiming@zededa.com>
this runs as an 'edgeview' container inside EVE, for access into EVE,
the Apps and it's site routable Apps
- it offers EVE side log search, network related debugging,
  system related debugging and show pubsub data
- it offers copying files from device, such as log files, onto e.g. operator's laptop
- it offers 'virtual port mapping' which allows TCP services into device or apps
- this patch does not include the session token/dispatcher ip,port privisionning,
  for which one can use e.g. https://github.com/lf-edge/eve/pull/2269 to build a
  private image and use 'zcli' to supply those configures for testing
- with the patch, one can build a client side 'edge-view' container for query
  or local virtual-portmapping endpoints, the container can also be used
  for ssh-mode access into the EVE for device debugging
- this patch has the Golang example program for 'edge-view' dispatcher which
  is needed for non-ssh mode with device running behind fw/nat/lte/proxy
- going to generate an EVE wiki doc to describe 'edge-view' in more detail

milan-zededa commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,451 @@
+			helpExample("log/Clock -type app", "display previous 30 minutes log contains 'Clock' in app log", false)
+			helpExample("log/certificate -time 2021-08-15T23:15:29Z-2021-08-15T22:45:00Z -json",
+				"display log during the specified time in RFC3339 format which contains 'certificate' in json format", false)
+		default:
```
I think a command showing reboot-reason and reboot-stack would be also helpful.

naiming-zededa replied:
i could not find where to look for this on the device, @milan-zededa do you know where it is saved?

milan-zededa replied:
They get stored under `/persist/reboot-reason` and `/persist/reboot-stack`.

milan-zededa commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,451 @@
+		case "url":
+			helpOn("url", "display url metrics for zedclient, zedagent, downloader and loguploader")
+		case "wireless":
+			helpOn("wireless", "display the iwconfig wlan0 info and wpa_supplicant.conf content")
```
Btw. Input to wwan service (config) is `/run/wwan/config.json` and the output from wwan service are `/run/wwan/status.json` and `/run/wwan/metrics.json`. Might be helpful to include that in the command output as well.

naiming-zededa replied:
sure.

milan-zededa reviewed (commented):
This is going to be super helpful for debugging. Really looking forward to this!

milan-zededa commented:
Maybe we could also use edge-view to expose profiling data and the current stack trace of pillar microservices with the help of the [pprof package](https://pkg.go.dev/net/http/pprof) (pprof would be enabled in microservices end exposed via http if some config option is enabled). WDYT?

giggsoff commented on pkg/edgeview/dispatcher/wss-server.go:
```diff
@@ -0,0 +1,265 @@
+)
+
+var upgrader = websocket.Upgrader{} // use default options
+// reqAddrTokeConn indexed by 'toke' then 'remoteAddr' strings
```
toke -> token

naiming-zededa replied:
ok.

giggsoff commented on pkg/edgeview/dispatcher/wss-server.go:
```diff
@@ -0,0 +1,265 @@
+package main
```
Seems, we need copyright in the first lines.

naiming-zededa replied:
missed that. will add.

giggsoff commented on pkg/edgeview/dispatcher/wss-server.go:
```diff
@@ -0,0 +1,265 @@
+func getOutboundIP() string {
+    conn, err := net.Dial("udp", "8.8.8.8:80")
+    if err != nil {
+        log.Fatal(err)
```
In case of connectivity issues seems reasonable to try again instead of fatal here.

naiming-zededa replied:
sure.

giggsoff commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,451 @@
+	"strings"
+	"time"
+
+	"golang.org/x/crypto/ssh"
```
Cannot find it in go.mod

naiming-zededa replied:
it has "golang.org/x/cryptio", i guess that covers this ssh

gkodali-zededa commented:
It will he helpful to add command line arguments related help information in the Readme file.

gkodali-zededa commented on images/docker-compose.yml.in:
```diff
@@ -52,6 +52,18 @@ services:
+            - etc:/etc:ro
+            - proc:/host/proc:ro
+        network_mode: host
+        #networks:
```
might as well remove this commented line.

naiming-zededa replied:
sure.

gkodali-zededa commented on pkg/edgeview/dispatcher/wss-server.go:
```diff
@@ -0,0 +1,265 @@
+    }
+    defer conn.Close()
+
+	if _, ok := r.Header["X-Session-Token"]; !ok {
```
This function needs indentation fixing.

naiming-zededa replied:
> Maybe we could also use edge-view to expose profiling data and the current stack trace of pillar microservices with the help of the [pprof package](https://pkg.go.dev/net/http/pprof) (pprof would be enabled in microservices end exposed via http if some config option is enabled). WDYT?

Good idea. I can add a story on this.

naiming-zededa replied:
@milan-zededa i realized we can do it right now, no story is needed. I just tried this:
1) compile gopprof into zedbox.go and listening on port 6060
2) with edge-view, do tcp/localhost:6060
3) another window on the laptop, get the file:

$ curl http://localhost:9001/debug/pprof/heap --output heap.file
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  100k    0  100k    0     0   363k      0 --:--:-- --:--:-- --:--:--  363k


$ go tool pprof heap.file
File: zedbox
Type: inuse_space
Time: Dec 10, 2021 at 1:51pm (PST)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof)
(pprof) top
Showing nodes accounting for 8941.76kB, 61.33% of 14580.73kB total
Showing top 10 nodes out of 175
      flat  flat%   sum%        cum   cum%
 1722.14kB 11.81% 11.81%  1722.14kB 11.81%  github.com/google/gopacket/layers.init
 1542.33kB 10.58% 22.39%  1542.33kB 10.58%  encoding/json.Marshal
 1536.14kB 10.54% 32.92%  1536.14kB 10.54%  google.golang.org/protobuf/internal/impl.fieldInfoForScalar
 1025.44kB  7.03% 39.96%  1025.44kB  7.03%  encoding/pem.Decode
  526.13kB  3.61% 43.57%  1038.13kB  7.12%  sync.(*Map).LoadOrStore
  522.70kB  3.58% 47.15%   522.70kB  3.58%  regexp/syntax.(*compiler).inst
  522.06kB  3.58% 50.73%   522.06kB  3.58%  github.com/lf-edge/eve/pkg/pillar/cmd/nodeagent.parseSMARTData.func1
  515.19kB  3.53% 54.26%   515.19kB  3.53%  regexp.onePassCopy
     515kB  3.53% 57.80%      515kB  3.53%  google.golang.org/protobuf/internal/filedesc.(*File).initDecls
  514.63kB  3.53% 61.33%   514.63kB  3.53%  github.com/lf-edge/eve/pkg/pillar/types.(*ConfigItemSpecMap).AddIntItem

eriknordmark replied:
> * compile gopprof into zedbox.go and listening on port 6060

@milan-zededa If we add this we should also add a rule to block port 6060 coming in on ethN ports.
@naiming-zededa do you automatically block any ports used between the edgeview container and EVE-OS or app instances so that they can not be accessed when coming in directly on ethN?

naiming-zededa replied:
From 'edge-view', no it does not install any ACLs. The above only shows if the EVE special image is coded to listen on this port (outside of edge-view domain), using edge-view 'tcp' option (it does not need to add anything for this gopprof usage), a remote laptop can access that gopprof service just like naturally routed into the device. But EVE can add any ACL rules when it offers this kind of service.

eriknordmark replied:
But in this example of goprof that http server has no security what so ever so we'd have to block it.

Maybe we should have EVE-OS have ACLs to drop all inbound TCP/UDP on the physical ports unless there is a portmap for an application so that we don't accidentally open up holes which can be accessed from the LAN.

naiming-zededa replied:
Also in this particular case, the `pprof` is listening on 'localhost:6060', so only edge-view can get access to that besides ssh into the device. So this is ok.

eriknordmark replied:
It would be good to check that the Linux kernel would block packets arriving on eth0 destined to 127.0.0.1:6060 (and ::1:6060 for IPv6). I suspect this is in some bogon-type check but we'd really depend on that being there for this pprof and other local port usage.

naiming-zededa replied:
@eriknordmark this is certainly the case, if the kernel does not listen on the ip:port combination, no ip service can be allowed in.

As in this case on my device:
the localhost:4822 is opened by 'edge-view' on my device:
07ab2c40-408a-4bc2-b9f2-8ca94235074f:~# netstat -ltnp
Active Internet connections (only servers)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      0 10.1.0.1:53             0.0.0.0:*               LISTEN      3664/dnsmasq
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN      1156/sshd -D -e [li
tcp        0      0 0.0.0.0:4822            0.0.0.0:*               LISTEN      1237/guacd
tcp        0      0 127.0.0.1:8234          0.0.0.0:*               LISTEN      2108/edge-view

1) i can telnet to it if I'm on the device:
07ab2c40-408a-4bc2-b9f2-8ca94235074f:~# telnet 127.0.0.1:8234
Connected to 127.0.0.1:8234

2) but remotely to the eth0, will be denied:
07ab2c40-408a-4bc2-b9f2-8ca94235074f:~# ip -4 a
7: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet 192.168.86.36/24 brd 192.168.86.255 scope global dynamic noprefixroute eth0
       valid_lft 73548sec preferred_lft 59083sec

% telnet 192.168.86.36:8234
192.168.86.36:8234: nodename nor servname provided, or not known

eriknordmark replied:
I was referring to having an attacker which telnets to 127.0.0.1:8234 with a TCP/IP stack and ARP entry which sends the TCP SYN to the attacked node. Question is whether Linux drops packets arriving on non-lo0 destined to 127.0.0.1.

naiming-zededa replied:
I'm not sure how one can achieve such a task. But this issue is generic to Linux and not specific to EVE. As in kubernetes, the API server listens on 127.0.0.1:6443, and it's not reachable from external to control the apps inside unless some proxy service is enabled on the host to bridge that.

naiming-zededa replied:
one reference: https://books.google.com/books?id=_zhCEAAAQBAJ&pg=PT68&lpg=PT68&dq=does+linux+access+packet+from+external+to+127.0.0.1?&source=bl&ots=HYAfEmC63N&sig=ACfU3U0QKNoq3OOL-uypanEjyWGE6kL4PA&hl=en&sa=X&ved=2ahUKEwinl-68gPf3AhXTno4IHVVRC8sQ6AF6BAgbEAM#v=onepage&q=does%20linux%20access%20packet%20from%20external%20to%20127.0.0.1%3F&f=false

it says: "It is possible to spoof such a packet by writing a fake source address (i.e. spoofing that a packet has a source and destination address of 127.0.0.1). Linux will normally filter such a packet when it arrives at an external interface"

gkodali-zededa commented on pkg/edgeview/dispatcher/wss-server.go:
```diff
@@ -0,0 +1,265 @@
+// debug set
+var needDebug bool
+
+func socketHandler(w http.ResponseWriter, r *http.Request) {
```
Can you please describe in comments how the hand shake happens between edge-node and edge-view client connections?

naiming-zededa replied:
will do.

gkodali-zededa commented on pkg/edgeview/dispatcher/wss-server.go:
```diff
@@ -0,0 +1,265 @@
+		}
+		return
+	}
+	token := r.Header["X-Session-Token"][0]
```
Is this token the binding factor between edge-view client and edge-node?

naiming-zededa replied:
yes, the token is the only thing to bind the sessions.

gkodali-zededa commented on pkg/edgeview/src/copyfile.go:
```diff
@@ -0,0 +1,210 @@
+		for {
+			select {
+			case message := <-copyMsgChn:
+				if !strings.Contains(string(message), startCopyMessage) {
```
Shouldn't these global variables copyMsgChn, startCopyMessage be defined in the current file instead of edge-view.go? It is a bit hard to read with usage and definitions in different files. Made me think that the file copy starts without the receiver acknowledging receipt of file information first.

naiming-zededa replied:
I was debating on to group them in one place or spreading them across. will think about this more.

gkodali-zededa commented on pkg/edgeview/src/copyfile.go:
```diff
@@ -0,0 +1,210 @@
+			fmt.Printf("file read error %v\n", err)
+			return
+		}
+		_ = websocketConn.WriteMessage(websocket.BinaryMessage, buffer[:n])
```
Should we not check for error and terminate the file transfer? For example when the client sends RST.

naiming-zededa replied:
yes, we can do that.

gkodali-zededa commented:
I think it's fine to print the client side logs to STDOUT.
Should the edge-node side container logs be plumbed into memlogd?

rvs commented on images/rootfs.yml.in:
```diff
@@ -26,6 +26,10 @@ onboot:
    - name: storage-init
      image: STORAGE_INIT_TAG
 services:
+   - name: edgeview
```
Hey @naiming-zededa -- I don't think including edge-view in the default build is a good idea. Let's discuss separately how can we have an architecture where this container can only be deployed via controller.

naiming-zededa replied:
@rvs  The container although it's running, but it does not do anything, unless the controller sends the command to assign token, ip:port, timeout, then the script inside the container will launch the real 'edge-view' program to run. And the script will kill it if the timeout expires. Will this be ok?

rvs commented on pkg/debug/ssh.sh:
```diff
@@ -7,4 +7,8 @@ echo -1 > /proc/sys/kernel/perf_event_paranoid
 
+mkdir -p /run/debug/usr/bin
+cp /usr/bin/lshw /run/debug/usr/bin/.
+cp /usr/bin/spec.sh /run/debug/usr/bin/.
```
We should either move this into whatever container you need them to be in or come up with some kind of a mounting solution -- this kind of copying shouldn't happen (it is OK for a WIP but not for when we need to merge this).

naiming-zededa replied:
I was thinking if the 'edge-view' is really used, and if we still keep the 'debug' container, then this allows us not to duplicate the two sets of services, if we later remove the 'debug' container, then those two can be moved into 'edge-view' container. but i can drop those to services for now.

rvs commented on pkg/dom0-ztools/rootfs/etc/init.d/010-eve-cgroup:
```diff
@@ -61,7 +61,7 @@ if [ -z "${ctrd_cgroup_cpus_limit}" ]; then
 
 CGROUPS="cpuset cpu cpuacct blkio memory devices freezer net_cls perf_event net_prio hugetlb pids systemd "
-EVESRVICES="sshd wwan wlan lisp guacd pillar vtpm watchdog xen-tools "
+EVESRVICES="sshd edgeview wwan wlan lisp guacd pillar vtpm watchdog xen-tools "
```
This should also NOT be needed if we make this edgeview container loadable on demand.

rvs commented on pkg/edgeview/Dockerfile.query:
```diff
@@ -0,0 +1,22 @@
+FROM lfedge/eve-alpine:6.7.0 as build
```
Why do we need two separate Dockerfiles? Can't we conditionalize one somehow?

naiming-zededa replied:
the 2nd one is not for EVE device, it's for someone to cd to the directory and make a 'client' docker container.

rvs commented on pkg/edgeview/Makefile:
```diff
@@ -0,0 +1,13 @@
+# build the websocket server/dispatcher, should compile on a machine in the
+# same architecture as the server runs the websocket dispather
+#
+wss-server:
```
Wait, so we really need 3 containers here correct?
  * wss-server
  * client
  * the one that gets deployed on EVE

naiming-zededa replied:
@rvs for EVE device, only the last one is needed. but the end-to-end service (besides the controller UI piece) the client and wss-server are needed, here I just offer a way to compile them and to use the same repo

eriknordmark replied:
@naiming-zededa when we build the one which gets deployed on EVE I assume it is built for Alpine i.e., busybox environment. Are the other two built as native for the built system? For instance, can one build client for Ubuntu and for Windows?

naiming-zededa replied:
@eriknordmark the EVE and 'client' are basically the same docker image both with
- FROM lfedge/eve-alpine:6.7.0 as build

the 'dispatcher' is an example code there, currently don't have docker wrapper around, i just build on an ubuntu machine and copy the binary to the dispatcher server (so it does not need to have all the docker infra).

rvs commented on pkg/edgeview/src/edge-view-init.sh:
```diff
@@ -0,0 +1,28 @@
+      echo "edge-view killed"
+    fi
+  fi
+done
```
This seems a tad complex -- why all this business with multiple entry points? Why not a simple "live-forever" loop?

naiming-zededa replied:
I thought this was a 'live-forever' loop. it does not exit.

rvs reviewed (changes requested):
Ok, I'll review go implementation separately (as I'm sure will @eriknordmark) but for now I'd like to focus on how this is going to be integrated with EVE. As I said -- I feel that bundling this into EVE by default is a bad idea from a security standpoint. Hence we need to make sure this is only deployed by the controller (the plumbing maybe different of course).

Let's discuss this part first.

rvs reviewed (commented):
In fact @naiming-zededa -- I'd like to propose that you split your PR into two PRs:
  1. simply add pkg/edge-view
  2. integration of pkg/edge-view into EVE

That way we can get #1 in very quickly and then spend some time focused on #2. Will this work for you?

naiming-zededa commented on pkg/edgeview/README.md:
```diff
@@ -0,0 +1,27 @@
+
+## ./dispatcher directory
+
+It containers an example of golang source for 'edgeview' websocket dispatcher running in the cloud or in some VPN servers
```
ok

naiming-zededa commented:
> I think it's fine to print the client side logs to STDOUT. Should the edge-node side container logs be plumbed into memlogd?

AFAIK, the edge-node side output is redirected into the memlogd. I got all the output in the newlog already.

naiming-zededa commented:
I committed the patch for the current review comments, please review:
- fixed some typo, copyright, indentions
- added 'lastreboot' option to display the reboot reasons and reboot-stack
- added wwan output to the 'wireless' option
- added retry for wss-server dispatcher to find local intf it uses
- moved some const and global vars to specific service files
- added comments on how user/edge-node to dispatcher connection relations in wss-server.go
- remove the edgeview in rootfs.yml.in, EVE build will NOT include edgeview container
- added images/rootfs-edgeview-kvm.yml file, to special build an eve image with edgeview, needs to do 'make HV=edgeview-kvm eve'; if we need something more fancy, we can do more optimization later. But now image build for EVE does NOT include edgeview
- removed copy spec.sh and lshw inside the debug container; added spec.sh and addpackage for lshw in edgeview
- added help option in README.md
- remove the client certs code in edgeview and dispatcher
-

naiming-zededa commented:
> In fact @naiming-zededa -- I'd like to propose that you split your PR into two PRs:
> 
>     1. simply add pkg/edge-view
> 
>     2. integration of pkg/edge-view into EVE
> 
> 
> That way we can get #1 in very quickly and then spend some time focused on #2. Will this work for you?

I have removed build edgeview into EVE. but added an build option for special image. we can do more optimization later if needed.

eriknordmark commented on pkg/edgeview/src/network.go:
```diff
@@ -0,0 +1,1052 @@
+	} else {
+		timeValue = "60"
+	}
+	cmd := "timeout " + timeValue + " tcpdump -i " + intf + " " + subs[1] + " -c 100"
```
Note that this arbitrary subs[1] argument allows the user to pass in e.g., "; reboot". The same is actually true for the intf argument - "tcpdump/eth1;reboot" will result in a reboot.

If we really want to restrict what the user can do we need a different approach with only specific arguments and those are checked to not include characters which are special to the shell, OR we avoid using the shell and instead parse the arguments and construct the argv array and pass that to execv(). Latter might be a reasonable compromise to reduce the amount of mischief which can be done using the shell.

eriknordmark replied:
@naiming-zededa can you check that intf and subs[1] does not contain anything but letters and digits? Othwewise I can use tcpdump to do any shell command.

naiming-zededa replied:
added the check for against all the up and valid interface name, and check for ';' string in the sub[1] in new commit

naiming-zededa replied:
@eriknordmark  committed the code to check 'intf' against all the 'UP' interfaces name, and check for ';' string in subs[1]

eriknordmark replied:
But ';' is only one of lots of characters which can be used to fool the shell. '|' is another one. Hence the request to only allow alphanumeric characters.

Basically all calls to runCmd need to have anything which can be possible controllerd by the user sanitized.
And to make this easier to review it would be a good idea to replace all of the unnecessary runcmd which invoke ls, cat, ip use the golang functions (ReadFile, Readdir, netlink) so that one can look more carefully at the ones which invoke tcpdump, ping, nslookup, grep, etc.

naiming-zededa replied:
@eriknordmark ok. but since this 'edge-view' also have the mode for 'ssh' to gather the data from device, that part can not be in golang functions. let me go through the list, and remove certain options (e.g. remove the above sub[1] completely) to have commands with less flexibility and more security way.

eriknordmark replied:
But as I asked in the other comment from today, isn't this ssh to handle old devices where EVE-OS has no edge-view support? 
While that might be useful to have in our back pocket for the cases where we have ssh access, it does not have to be part of the Edge-view which we put in EVE-OS; there we will run the commands on the device itself and will access the files and directories etc.

naiming-zededa replied:
I can remove that, then the ssh-mode can not be supported in the future unless we keep different repos, currently e.g. Ruslan's test job for traffic is using the 'edge-view' ssh-mode.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,463 @@
+			retStr = string(retBytes)
+		}
+	} else {
+		retStr, err = remoteRun("root", evDevice, sshprivkey, cmd)
```
When is this ssh invocation used? Is this code used on the client in addition to being used in EVE? Confused.

naiming-zededa replied:
edge-view has two modes, one is directly using ssh into the device, without the dispatcher thing; the other is the using dispatcher. they share most of the code, only when executing the commands, one is the launch the 'ssh' as in here, the other is above the, which is 'edge-view' "run-on-server" which is local.

naiming-zededa replied:
and this 'ssh' is not the 'ssh' we were discussing ;-)

eriknordmark replied:
I wonder if we can structure or comment the code to make the different modes more clear.

naiming-zededa replied:
made extra comments in this section.

naiming-zededa replied:
added comments.

eriknordmark replied:
Wait, what does the ssh to the server assume here? That ssh as root is enabled into the edge-node? That might be useful for old devices which are not running edge-view, but I don't see why we need it going forward.
So do we need to include this in the edge-view code we integrate?

naiming-zededa commented:
Committed more changes to this PR:
- Add JWT handling for token
- Add Hmac authentication with JWT's nonce for all messages
   between the device and user session
- Enforce simple edgeview policy for now
- Add show 'techsupport' cmd, which gathers number of
  cmd's output and download a compressed file to local computer
- Add 'copy-logfiles' option to specify a time range and download
  all the gzip'ed logfiles from the device to local computer, particular
  useful since now we skip app-log if app configured it out
- Publish edgeview status
- With shell command, use exec.Command(,args...) to prevent multiple
   commands
- add docker mount '/run/edgeview' as normal access, and keep '/run' as readonly
- etc.

naiming-zededa commented:
updated this edge-view PR:
- some bug fixes
- added multiple instance support for edge-view (allow multiple sessions into the same device at the same time)
- added encryption option in addition to the authentication
- track the user/client endpoint IP
- added ping/pong http handling on dispatcher for external health check later
- add edge-view cmd logging to event tag with special object-key/name (can be used in controller for formatting events)

rvs commented:
Hey @naiming-zededa -- I'd love to merge this ASAP, but there's a few things that may take longer. So... here's what I propose:
   * drop changes to `rootfs.yml.in` -- for now we'll still build this container as a one-off
   * drop `spec.sh` since we will be doing `debug` and `edge-view` merging as a separate PR
 
And finally, why do you need `mkdir -p /run/edgeview` in storage-init.sh? This is odd since none of the others need it, I guess.

Finally, please coordinate with @ruslan-zededa so that there's a LF Edge Docker HUB repo created for edgeview ahead of time. Otherwise the build on master will break.

naiming-zededa commented:
@rvs 

>     * drop changes to `rootfs.yml.in` -- for now we'll still build this container as a one-off
>     * drop `spec.sh` since we will be doing `debug` and `edge-view` merging as a separate PR

Done, just submitted the patch.

> And finally, why do you need `mkdir -p /run/edgeview` in storage-init.sh? This is odd since none of the others need it, I guess.

This part 'edge-view' is special, since it only volume mounts all with read-only, and it needs to publish stats and needs to separately mount '/run/edgeview' as 'read-write'. This is to make sure that directory is there during the container comes up in the process.

> Finally, please coordinate with @ruslan-zededa so that there's a LF Edge Docker HUB repo created for edgeview ahead of time. Otherwise the build on master will break.

Ok.

naiming-zededa commented:
new commit to:
1) address some of the comments
2) fix a bug in multiple instances, the JWT token string passed in was incorrect
3) make the websocket disconnect retry delay to be exponential back-off

ruslan-zededa commented:
@naiming-zededa repo  [eve-edgeview](https://hub.docker.com/r/lfedge/eve-edgeview)  created

naiming-zededa commented:
@rvs the docker repo https://hub.docker.com/r/lfedge/eve-edgeview is created by @ruslan-zededa already.

eriknordmark commented on pkg/edgeview/src/copyfile.go:
```diff
@@ -0,0 +1,359 @@
+		return
+	}
+	file := path[1]
+	info, err := os.Stat(file)
```
This SplitN is OK since it implicitly verifies that it is a valid filename. But I don't know if one can make that more explicit by calling filepath.Abs() on the name. From what I can tell that still doesn't check that it is a syntactically valid pathname so maybe just using Stat() is sufficient.

FWIW I'm looking at all the SplitN() calls in this PR.

eriknordmark commented on pkg/edgeview/src/edge-view.go:
```diff
@@ -0,0 +1,580 @@
+	if pqueryopt != "" {
+		if strings.HasPrefix(pqueryopt, "log/") {
+			logs := strings.SplitN(pqueryopt, "log/", 2)
+			logopt = logs[1]
```
Here logopt can be an arbitrary string. Where does it get used? Can we constrain it to be a valid filename syntax?

naiming-zededa replied:
the syntax 'log/<search-string>', is a string to narrow down the output in the log search. I'll put regex() for qualified string check.

eriknordmark commented on pkg/edgeview/src/network.go:
```diff
@@ -0,0 +1,1134 @@
+		if strings.Contains(opt, "/") {
+			items := strings.SplitN(opt, "/", 2)
+			opt = items[0]
+			substring = items[1]
```
This can contain arbitrary shell escapes. Can we constain the syntax of substring for the different network command?

naiming-zededa replied:
will do.

eriknordmark commented on pkg/edgeview/src/network.go:
```diff
@@ -0,0 +1,1134 @@
+			return
+		}
+	}
+	if subs[1] != "" && strings.Contains(subs[1], ";") {
+		fmt.Printf("tcpdump command can not have ';' in the option\n")
+		return
+	}
```
This check is a distraction since it doesn't prevent the range of shell magic.

eriknordmark commented on pkg/edgeview/src/system.go:
```diff
@@ -0,0 +1,766 @@
+	var item string
+	if strings.Contains(opt, "ps/") {
+		opts := strings.SplitN(opt, "ps/", 2)
+		item = opts[1]
```
Is this to look for a particular command string? If so can we validate that it is an alphanumeric string (a-z0-9 allowed)?

naiming-zededa replied:
will do.

eriknordmark commented on pkg/edgeview/src/system.go:
```diff
@@ -0,0 +1,766 @@
+	printColor(" - in "+dir, colorCYAN)
+	lines := splitLines(retStr)
+	for _, jfile := range lines {
+		retStr, err := runCmd("cd "+dir+" && find . -name "+jfile+" -print", false, false)
```
Using golang's Readdir to find the matching names makes a lot of sense to me.

naiming-zededa replied:
Right. But this program supports remote ssh queries. Maybe later on we stop supporting ssh into the device, we can convert then. I can make a comment on this.

eriknordmark commented on pkg/edgeview/src/system.go:
```diff
@@ -0,0 +1,766 @@
+	}
+}
+
+func runShell(opt string) {
+	shell := strings.SplitN(opt, "shell/", 2)
```
How often have you used this against EVE-OS? Can we drop this generic support and instead gather the list of commands which folks want and make them be specific commands?

eriknordmark commented on pkg/edgeview/src/tcp.go:
```diff
@@ -0,0 +1,723 @@
+
+func processTCPcmd(opt string, remotePorts map[int]int) (bool, int, map[int]int) {
+	tcpopts := strings.SplitN(opt, "tcp/", 2)
+	tcpparam := tcpopts[1]
```
What syntax do we need to support for tcpparam? Is it a well-known set of IP + port-related synax which we can check for validity.

naiming-zededa replied:
Yes, it can be checked at the top level, at runNetwork() loop for all the options strings. will do.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,608 @@
+
+func getHostname() string {
+	var hostname string
+	retStr, err := runCmd("hostname", false, false)
```
There is a golang function for this called os.Hostname()

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,608 @@
+	if err != nil {
+		return hostname
+	}
+	hostname = strings.TrimSuffix(retStr, "\n")
```
No need for this if you use os.Hostname()

naiming-zededa replied:
see above.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,608 @@
+		fmt.Printf("  EVE-OS release %s", retStr)
+	}
+	fmt.Printf("  Edge-View Ver: %s\n", edgeViewVersion)
+	retStr, err = runCmd("uptime", false, false)
```
"github.com/shirou/gopsutil/host"
 has a Info function which returns a struct with an Uptime field. See code in zedagent/reportinfo.go

naiming-zededa replied:
see above. Will make a comment on this.

naiming-zededa replied:
ok, if i remove the 'ssh-mode' option, i can change them all.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,608 @@
+
+// dynamically install package for uncommon options
+func addPackage(programName, pkgName string) error {
+	retStr, err := runCmd("which "+programName, false, false)
```
Can we not require the callers to know the exact pathname for programName so this check can just be an os.Stat() to check whether e.g., /usr/sbin/tcpdump exists?

naiming-zededa replied:
Ok, let me find out all the programs are and pass in the whole path to it.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,608 @@
+	if runOnServer {
+		// the program is running as a container on the device as a 'server',
+		// it can directly access the linux shell for requested command
+		retBytes, err = exec.Command("sh", "-c", cmd).Output()
```
This sh -c needs to go away. Far too much security risk.
The commands should be passed as []string from the caller of runCmd and be fed into os.Exec(). That removes many of the shell tricks which can be used to launch attacks.

naiming-zededa replied:
For some of the 'edge-view' operations, they have grep, awk, etc things, but i'll do this:
- all the user pass in params/strings, it has to be qualified by regex allowed characters check
- we use the 'sh', '-c' cmd, only for those internal operations we know there is no problem
- for others all use []string to the os.Exec() of golang side

naiming-zededa replied:
@eriknordmark yes, I have limited to 'sh', '-c' to the places only necessary for those places needs them, and converted all the others to the []string into the exec(); for all the user supplied string, has to pass the regex() allowed check; I have converted all the previous coded 'cat', 'ls' to using the golang functions now.

eriknordmark replied:
How about introducing specific commands for ls and cat so that we don't need the shell/ command?

naiming-zededa replied:
Will do. At least to start with.

naiming-zededa commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,608 @@
+
+	var uuidStr string
+	if isServer {
+		retStr, err := runCmd("cat /persist/status/uuid", false, false)
```
see above.

naiming-zededa commented:
new patch:
- removed the 'ssh-mode' from 'edge-view'
- converted most of the runCmd() into goland functions
- most for the os.Exec() uses the []string params
- for a few commands (mainly for internal ops) still use 'sh', make sure all the strings passed are sanitized

naiming-zededa commented:
submit this commit to make the same 'edge-view' container to be used for both server and client sides. we'll only have one docker repo for the image: lf-edge/eve-edgeview.

naiming-zededa commented:
add '-' and '_' to the allowed character for user input

naiming-zededa commented:
in the latest patch:
- removed 'shell' command
- removed the 'sh' '-c' for exec.Command
- added 'ls', 'cat', 'du', 'top' new commands

eriknordmark commented on pkg/edgeview/README.md:
```diff
@@ -0,0 +1,38 @@
+  pub/ [baseosmgr domainmgr downloader global loguploader newlogd nim nodeagent tpmmgr vaultmgr volumemgr watcher zedagent zedclient zedmanager zedrouter]
+
+  [acl app arp connectivity flow if mdns nslookup ping route socket speed tcp tcpdump trace url wireless]
+  [app configitem cp datastore download hw lastreboot model newlog pci ps cipher usb shell volume]
```
Please remove shell and add the new commands.

naiming-zededa replied:
ok.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,665 @@
+}
+
+func runCmdInFunction(cmdOpt string) ([]byte, error) {
+	shellcmd := strings.Fields(cmdOpt)
```
Nit on variables names, but since there is no shell any more it would make sense to rename the 30-odd occurances of shell and shellcmd to just cmd.

naiming-zededa replied:
will fix them all.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,665 @@
+		return "", fmt.Errorf("shell command invalid\n")
+	}
+
+	c1 := exec.Command(shellcmd1[0], shellcmd1[1:]...)
+	c2 := exec.Command(shellcmd2[0], shellcmd2[1:]...)
```
Does this work even when len(shellcmd1) == 1 i.e., does the [1::] become and empty slice?

eriknordmark commented on pkg/edgeview/src/system.go:
```diff
@@ -0,0 +1,1022 @@
+
+	for _, l := range files {
+		if strings.Contains(l.Name(), "pillar-panic-stack.") {
+			fields := strings.Fields(l.Name())
+			n := len(fields)
+			retbytes, err := ioutil.ReadFile("/persist/newlog/panicStacks/"+fields[n-1])
```
Does this need a check for n >=1 to avoid referencing fields[-1]?

naiming-zededa replied:
added a check.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,665 @@
+	return nil
+}
+
+func runCmd(cmd string, isPrint bool) (string, error) {
```
Currenly the caller need to create a string with whitespace and it is hard to read to see that there is whitespace between the different arguments. Why not make this funftion (and runCmdInFunction) take the same approach as exec.Command i.e., prog string, args []string. That reduces the risk of whitespace errors and the need for a Fields call inside runCmdInFunction.

It could be that some callers which variable number of args need to call string.Fields but then at least it is explicit which ones pass in more complex lists of arguments and which ones have a fixed number.

naiming-zededa replied:
changed to runCmdInFunction()

eriknordmark replied:
I don't see a change which relates to by comment to use []string instead of string.

naiming-zededa replied:
ok, i misunderstood. let me change.

eriknordmark commented on pkg/edgeview/src/edge-view.go:
```diff
@@ -0,0 +1,537 @@
+	if pqueryopt != "" {
+		if strings.HasPrefix(pqueryopt, "log/") {
+			logs := strings.SplitN(pqueryopt, "log/", 2)
+			logopt = logs[1]
```
Missing check for len(logs) >= 2 AFAICT

eriknordmark replied:
Or != 2 as you have elsewhere.

eriknordmark commented on pkg/edgeview/src/edge-view.go:
```diff
@@ -0,0 +1,537 @@
+			}
+		} else if strings.HasPrefix(pqueryopt, "pub/") {
+			pubs := strings.SplitN(pqueryopt, "pub/", 2)
+			ppubsubopt = pubs[1]
```
Ditto. A different approach would be to use TrimPrefix but your call.

eriknordmark commented on pkg/edgeview/src/network.go:
```diff
@@ -0,0 +1,1243 @@
+		var substring string
+		if strings.Contains(opt, "/") {
+			items := strings.SplitN(opt, "/", 2)
+			opt = items[0]
```
Also missing check for len(items) != 2 before using items[1]

eriknordmark commented on pkg/edgeview/src/network.go:
```diff
@@ -0,0 +1,1243 @@
+	}
+
+	subs := strings.SplitN(subStr, "/", 2)
+	intf := subs[0]
```
Also missing check on len; you use subs[1] further down.

eriknordmark commented on pkg/edgeview/src/system.go:
```diff
@@ -0,0 +1,1022 @@
+		return
+	}
+	if strings.Contains(opt, "ps/") {
+		opts := strings.SplitN(opt, "ps/", 2)
```
Missing len check before referencing [1]/

eriknordmark commented on pkg/edgeview/src/system.go:
```diff
@@ -0,0 +1,1022 @@
+		item = opts[1]
+	}
+
+	matches, _ := filepath.Glob("/proc/*/exe")
+	for i, file := range matches {
```
Lots of code to screen scrape /proc output. Why not use github.com/shirou/gopsutil/process which is what we use in domainmgr to extract information about the processes?

naiming-zededa replied:
right, github.com/shirou/gopsutil/process seems to get the process related stats, but I need to get more than that, memory, net routes, intf arp, conntrack stats, etc.

eriknordmark replied:
This function only gets info for the processes.

FWIW there is a netlink package we use to get routes, intf, etc.
I thought Gopi pushed up some fixes to a conntrac package as well a few years back. So we can avoid screen scaping output in many cases.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,665 @@
+	}
+
+	if strings.Contains(jdata.Dep, "/") {
+		urls := strings.SplitN(jdata.Dep, "/", 2)
```
Missing len != 2 check

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,665 @@
+			cont2s := strings.Split(conts[1], ".zededa.net")
+			if len(cont2s) == 2 {
+				cluster := cont2s[0]
+				controller := strings.Replace(serverurl, cluster, "\033[0;93m"+cluster+"\033[0m", 1)
```
What does this do adding the control characters? Color??

naiming-zededa replied:
yes, a non production cluster name is highlighted in color.

eriknordmark replied:
Wouldn't be a bad idea to add comments about the color codes. I see you have some color function but there is also code which adds control characters here and there.

naiming-zededa replied:
added a new func 'getColorStr()' to handle string replacement.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,665 @@
+func splitLines(inStr string) []string {
+	outStr := strings.Split(inStr, "\n")
+	n := len(outStr)
+	return outStr[:n-1]
```
Sure this n-1 can't reference index -1 which doesn't exist? AFAIK there is a function in the string package to do what you want.

eriknordmark commented on pkg/edgeview/src/log-search.go:
```diff
@@ -0,0 +1,282 @@
+	}
+}
+
+func getTimeSec(timeline string, now int64) (int64, int64) {
```
Looks like this function is extracting two timestamps. Is that correct? A comment would be helpful.

naiming-zededa replied:
will do.

eriknordmark commented on pkg/edgeview/src/log-search.go:
```diff
@@ -0,0 +1,282 @@
+	} else {
+		times := strings.Split(timeline, "-")
+		f1, err1 := strconv.ParseFloat(times[0], 16)
+		f2, err2 := strconv.ParseFloat(times[1], 16)
```
Missng check for len(times) before referencing [1]

eriknordmark commented on pkg/edgeview/src/network.go:
```diff
@@ -0,0 +1,1243 @@
+	}
+	pathname := ""
+	paths := strings.Split(path, "/")
+	if len(paths) > 3 {
```
What output does this produce when the len is shorter than 3?

naiming-zededa replied:
changed to len(paths) != 3, then bail out.

eriknordmark commented on pkg/edgeview/src/network.go:
```diff
@@ -0,0 +1,1243 @@
+	if err == nil {
+		printTitle(" wpa_supplicant.conf:", colorCYAN, false)
+		lines := strings.Split(string(retbytes), "\n")
+		for _, l := range lines[:len(lines)-1] {
```
CHeck len(lines) before using as an index.

eriknordmark commented on pkg/edgeview/src/system.go:
```diff
@@ -0,0 +1,1022 @@
+			matchStr = strings.Split(file, "*")
+			n := len(matchStr)
+			if n < 2 {
+				fmt.Printf("len %d, %v\n", len(matchStr), err)
```
Why log err - it is nil here.

eriknordmark commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,665 @@
+	}
+}
+
+func checkOpts(opt string, optslice []string) ([]string, error) {
```
What does this function check? Please add comment.

naiming-zededa replied:
added.

naiming-zededa commented on pkg/edgeview/src/pubsub.go:
```diff
@@ -0,0 +1,133 @@
+		if len(dir1) < 2 {
+			continue
+		}
+		paths := strings.Split(dir1[1], "/")
```
the previous split is on 'newdir + "/"', there can be further directories on the dir1[1] here.

eriknordmark replied:
Ah - less than obvious but ok.

naiming-zededa commented on pkg/edgeview/src/basics.go:
```diff
@@ -0,0 +1,665 @@
+	return opts, nil
+}
+
+func isValidOpt(value string, optslice []string) bool {
```
changed to isDefinedOpt()

eriknordmark commented on pkg/edgeview/src/network.go:
```diff
@@ -0,0 +1,1256 @@
+}
+
+func runAppACLTblAddr(op, tbl, ipaddr string) {
+	retStr, err := runCmd("iptables "+op+" -t "+tbl, false)
```
To examplify my point about runCmd, it is a lot less error prone to do
   args := []string{"op", "-t", tbl"}
  runCmd("iptables", args)
thus runCmds second argument is of type []string and runCmd does not do a strings.Fields().

naiming-zededa replied:
Ok, let me scan through this.

eriknordmark reviewed (approved):
Run eden tests
